pkg/sendconfig/common_workflows.go

```diff
@@ -57,8 +71,14 @@ func UpdateKongAdminSimple(ctx context.Context,
 		targetConfig, kongConfig.FilterTags, nil, lastConfigSHA, false,
 	)
 	if err != nil {
+		if diagnostic != (util.ConfigDumpDiagnostic{}) {
+			diagnostic.FailedConfigs <- *diagConfig
+		}
 		return nil, err
 	}
+	if diagnostic != (util.ConfigDumpDiagnostic{}) {
+		diagnostic.SuccessfulConfigs <- *diagConfig
```

Comment:
We have two channels being used here which are not buffered, and therefore will block for an unknown period of time in a performance sensitive area of the code.

It's common to see unbuffered (blocking) channels for reading data (e.g. `<-ctx.Done()` or `<-ticker.Tick.C`) as this indicates fairly clearly to the caller that they either need to wait for that component to finish, OR they need to do other work in a `select {` statement. The opposite (having unbuffered channels for writing) is a little bit less common and has different connotations.

At a personal level I prefer _not_ to expose unbuffered write channels for sending data to actors within a program in favor of using buffered channels and some kind of backpressure system that can start to fail and raise alarms if the component is unable to keep up with its work queue (without blocking other components). I suggest we make some changes so as to ensure processing diagnostics does not block the update of the Kong Admin API so that this performance sensitive code doesn't not have to wait for the diagnostic server to receive messages before work can continue.

There is prior art within our codebase for a server/actor that processes input data asynchronously from other components that employs a buffering and backpressure technique to avoid the problem in our `Proxy` interface implementation. I'm not trying to insinuate that I would _only_ accept such a solution but instead just wanted to present it for your considerations.

Let me know your thoughts on this, I don't want this to block for too long but I do think it's worth a discussion.

Reply:
This is reasonable. Initial implementation lacked it for my inexperience with channels and not seeing any issues with it perceptibly blocking so long as the recipient was running. That definitely will depend on whether the recipient goroutine is actively running though, and even though the CPU wait for that likely isn't long, it is _there_. 98cacd13e4bfd8bfc9d3650fbf99ccd2437943eb